tree/treeplayer/src/TArrowDS.cxx

```diff
@@ -0,0 +1,324 @@
+   auto verifyColumnSize = [&table = fTable](std::shared_ptr<arrow::Column> column, int nRecords)
+   {
+      if (column->length() != nRecords) {
+         std::string msg = "Column ";
+         msg += column->name() + " has a different number of entries.";
+         throw std::runtime_error(msg);
+      }
+   };
+
+   /// For the moment we support only a few native types.
+   auto verifyColumnType = [](std::shared_ptr<arrow::Column> column) {
+      auto verifyType = std::make_unique<VerifyValidColumnType>();
+      auto result = column->type()->Accept(verifyType.get());
+      if (result.ok() == false) {
+         std::string msg = "Column ";
+         msg += column->name() + " contains an unsupported type.";
+         throw std::runtime_error(msg);
+      }
+   };
+
+   /// This is used to create an index between the columnId
+   /// and the associated getter.
+   auto addColumnToGetterIndex = [&index = fGetterIndex](int columnId)
+   {
+      index.push_back(std::make_pair(columnId, index.size()));
+   };
+
+   /// Assuming we can get called more than once, we need to
+   /// reset the getter index each time.
+   auto resetGetterIndex = [&index = fGetterIndex]() { index.clear(); };
+
+   /// This is what initialization actually does
+   filterWantedColumns();
+   resetGetterIndex();
+   auto nRecords = getRecordsFirstColumn();
+   for (auto &columnName : fColumnNames) {
+      auto columnIdx = fTable->schema()->GetFieldIndex(columnName);
+      addColumnToGetterIndex(columnIdx);
+
+      auto column = fTable->column(columnIdx);
+      verifyColumnSize(column, nRecords);
+      verifyColumnType(column);
+   }
+   SetNSlots(fNSlots);
+}
+
+////////////////////////////////////////////////////////////////////////
+/// Destructor.
+TArrowDS::~TArrowDS()
+{
+}
+
+// FIXME: this is slow. One could easily cache results, but for
+// now we do not care doing it.
+const std::vector<std::string> &TArrowDS::GetColumnNames() const
+{
+   return fColumnNames;
+}
+
+std::vector<std::pair<ULong64_t, ULong64_t>> TArrowDS::GetEntryRanges()
+{
+   auto entryRanges(std::move(fEntryRanges)); // empty fEntryRanges
+   return entryRanges;
+}
+
+std::string TArrowDS::GetTypeName(std::string_view colName) const
+{
+   auto field = fTable->schema()->GetFieldByName(std::string(colName));
+   if (!field) {
+      std::string msg = "The dataset does not have column ";
+      msg += colName;
+      throw std::runtime_error(msg);
+   }
+   TDFTypeNameGetter typeGetter;
+   auto status = field->type()->Accept(&typeGetter);
+   if (status.ok() == false) {
+      std::string msg = "TArrowDS does not support a column of type ";
+      msg += field->type()->name();
+      throw std::runtime_error(msg);
+   }
+   return typeGetter.result();
+}
+
+bool TArrowDS::HasColumn(std::string_view colName) const
+{
+   auto field = fTable->schema()->GetFieldByName(std::string(colName));
+   if (!field) {
+      return false;
+   }
+   return true;
```

Comment:
Why expose the fact that `field` is a `std::shared_ptr`? This would work even in the case the API changed. I can change it, but again I think it's a matter of taste.

Reply:
It's only acknowledging that field is a pointer type, and that it should not be null, but you could also use `return !!field;` to avoid the if (which was the intent of my first comment) without exposing the type of `field`.